File: Problem 3/Solution 1.dart
import 'dart:math';

BigInt nextFactor(BigInt num) {
  var next = num + BigInt.one; //, limit = sqrt(num.toDouble()).ceil();

  if (!checkIfPrimeNumber(next)) next = next + BigInt.one;

  return next;
}

bool isPrime(BigInt num) {
  var limit = sqrt(num.toDouble()).ceil();

  if ((num).remainder(BigInt.two) == 0) return false;
  if ((num).remainder(BigInt.from(3)) == 0) return false;
  if ((num).remainder(BigInt.from(5)) == 0) return false;
  if ((num).remainder(BigInt.from(7)) == 0) return false;

  for (var i = 1; i < limit; i++) if (num % BigInt.from(i) == 0) return false;

  return true;
}

bool checkIfPrimeNumber(BigInt number) {
  var limit = sqrt(number.toDouble()).ceil();

  if (number == 2 || number == 3) {
    return true;
  } else if (number <= BigInt.one ||
      number % BigInt.two == 0 ||
      number % BigInt.from(3) == 0) {
    return false;
  }
  for (int i = 5; i * i <= limit; i += 6) {
    if (number % BigInt.from(i) == 0 || number % (BigInt.from(i + 2)) == 0) {
      return false;
    }
  }
  return true;
}

void main(List<String> args) {
  BigInt num = BigInt.parse('600851475143'),
      largestPrimeFactor = BigInt.two,
      primeNum = BigInt.two;

  while (num != BigInt.one && primeNum <= num) {
    if (num % primeNum == BigInt.zero) {
      num = num ~/ primeNum;
      largestPrimeFactor = primeNum;
      // primeNum = BigInt.two;
      print('$num : $largestPrimeFactor');
    } else
      primeNum = nextFactor(primeNum);

    // print(largest_prime_factor);
  }

  print(largestPrimeFactor);
}

File: Problem 3/Solution 1.cpp
#include <iostream>
#include <cmath>

template< typename T>

void print(T any) {
    std::cout << any << std::endl;
}

int nextFactor(long long num) {
    long long next = num++;
    long long limit = ceil(sqrt(next));

    for (int i = 1; i <= ceil(limit); i++) {
        if (next % i == 0) {
            next++;
            limit = ceil(sqrt(next));
            i = 1;
        }
    }

    return next;
}

int main(int argc, char const* argv[])
{
    long long num = 600851475143;
    int largest_prime_factor = 2;

    while (num != 1) {
        if (num % largest_prime_factor == 0)
            num /= largest_prime_factor;
        else
            largest_prime_factor = nextFactor(largest_prime_factor);
    }

    print(largest_prime_factor);
}
